KAFKA-19588: Reduce waiting for event completion in AsyncKafkaConsumer.poll()

kirktrue opened the pull request:
We create—and wait on—PollEvent in Consumer.poll() to ensure we wait for
reconciliation and/or auto-commit. However, reconciliation is relatively
rare, and auto-commit only happens every N seconds, so the remainder of
the time, we should try to avoid sending poll events.

kirktrue committed:
KAFKA-19588: Reduce number of events generated in AsyncKafkaConsumer.poll()

We create—and wait on—PollEvent in Consumer.poll() to ensure we wait for reconciliation and/or auto-commit. However, reconciliation is relatively rare, and auto-commit only happens every N seconds, so the remainder of the time, we should try to avoid sending poll events.

kirktrue committed:
Updates

kirktrue committed:
Updated naming

kirktrue committed:
Made AutoCommitState a top-level class and refactoring to support

kirktrue committed:
[WIP] Clean up

kirktrue committed:
Revert changes to CommitRequestManager to fix failures in CommitRequestManagerTest

kirktrue committed:
Merge branch 'trunk' into KAFKA-19588-avoid-poll-event

kirktrue commented:
@lianetm Could you add the `ci-approved` label so that I can see how this runs with GitHub Actions? Thanks!

kirktrue committed:
Merge branch 'trunk' into KAFKA-19588-avoid-poll-event

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -859,16 +865,19 @@ public ConsumerRecords<K, V> poll(final Duration timeout) {
+            // the interval time or reconciling new assignments
+            applicationEventHandler.add(event);
+
+            if (reconciliationInProgress.get() || autoCommitState.shouldAutoCommit()) {
```
Couldn't we end up with a race condition here if the app thread sees `autoCommitState.shouldAutoCommit()` false at this point (because interval hasn't expired just yet), but by the time the background checks the same when processing the poll event the interval expired?

In that case, I expect the background would trigger the auto-commit while the app thread moved onto updating positions for fetching (and that leads to a whole new set of race conditions that we already dealt with before). Basically, whatever change we introduce here to not wait on Poll, needs to ensure that we retrieve the positions to commit before moving on to update fetch positions, that's the main challenge with this change I expect. Thinking, but not sure yet how to address that if we don't wait on Poll. Thougts?

AndrewJSchofield replied:
I agree with @lianetm that this is opening up the risk of race conditions. However, I think the principle here is a good one. The risk part here is related to the auto-commit timer. If auto-commit is not enabled, we absolutely know that we are not racing with the auto-commit timer. If it is enabled, we are potentially in a race. So, a slight twist on this can safely optimise when auto-commit is not enabled.

lianetm replied:
Agreed, and trying to push it a bit more, even with auto-commit enabled, maybe we can still optimise if we check the timer in the app thread, and then use an atomic var "timerExpired" that the background can reuse (instead of checking the timer again). 
So we would optimse if auto-commit disabled, or if it's enabled but the timer hasn't expired.

kirktrue replied:
> Couldn't we end up with a race condition here if the app thread sees `autoCommitState.shouldAutoCommit()` false at this point (because interval hasn't expired just yet), but by the time the background checks the same when processing the poll event the interval expired?

No, simply because nothing updates the timer's sense of the current time between when the application thread submits the `PollEvent` and when it's processed by the network thread.

The application thread determines the timestamp used by the auto-commit timer via `PollEvent` (and `AssignmentChangeEvent`). And it's only checked when `CommitRequestManager.updateTimerAndMaybeCommit()` is invoked by the `ApplicationEventProcessor`. So when we update the timer in the application thread via `AutoCommitState.updateTimer()` and then call `AutoCommitState.shouldAutoCommit()`, we should get the same result in either thread, at least until the auto-commit is actually kicked off.

lianetm reviewed (commented):
Thanks @kirktrue , took a first look, one high level concern for now

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AutoCommitState.java:
```diff
@@ -0,0 +1,200 @@
+     * Reset the auto-commit timer to the provided time (backoff), so that the next auto-commit is
+     * sent out then. If auto-commit is disabled this will perform no action.
+     */
+    void resetTimer(long retryBackoffMs);
```
nit: Why not `final long retryBackoffMs` also?

kirktrue replied:
Done.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AutoCommitState.java:
```diff
@@ -0,0 +1,200 @@
+            this.log = logContext.logger(AutoCommitState.class);
+            this.timer = time.timer(autoCommitInterval);
+            this.autoCommitInterval = autoCommitInterval;
+            this.hasInflightCommit = new AtomicBoolean();
```
Doesn't the presence of `synchronized` on all of these methods make the use of `AtomicBoolean` redundant?

kirktrue replied:
Yes. Fixed.

AndrewJSchofield reviewed (changes requested):
Just a couple of trivial initial comments. I've skimmed over the PR and understand the overall flow now. I'll do a more in-depth review shortly.

kirktrue committed:
Merge branch 'trunk' into KAFKA-19588-avoid-poll-event

kirktrue committed:
Merge branch 'trunk' into KAFKA-19588-avoid-poll-event

kirktrue committed:
Remove redundant autoCommitState timer update in poll

kirktrue committed:
Updates for clarity

kirktrue committed:
Merge branch 'trunk' into KAFKA-19588-avoid-poll-event

kirktrue committed:
Refactoring to make the intention a little more obvious and in keeping with the style from KAFKA-19589

kirktrue committed:
Reverting unnecessary whitespace diffs

kirktrue commented:
@lianetm @AndrewJSchofield—Could you please add the `ci-approved' label? Thanks!

kirktrue commented:
@chia7712 @frankvicky—would you or someone you know to be familiar with the new `AsyncKafkaConsumer` be willing to help review this PR? Thanks!

frankvicky commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java:
```diff
@@ -140,7 +145,7 @@ public abstract class AbstractMembershipManager<R extends AbstractResponse> impl
      * assignmentReadyToReconcile. This will be true if {@link #maybeReconcile(boolean)} has been triggered
      * after receiving a heartbeat response, or a metadata update.
      */
-    private boolean reconciliationInProgress;
+    private final SharedReconciliationState reconciliationState;
```
Should we also update the Javadoc of this variable?

kirktrue replied:
I updated the JavaDoc for `reconciliationState` (and `autoCommitState`) to be more up-to-date and clear. Thanks!

frankvicky reviewed (commented):
@kirktrue: Thanks for the PR.
I have taken a quick pass and left a comment.
I will deep dive shortly.

kirktrue committed:
Merge branch 'trunk' into KAFKA-19588-avoid-poll-event

kirktrue committed:
Optimization to send only one PollEvent per poll() does not work, need to send one for each loop

kirktrue committed:
Merge branch 'trunk' into KAFKA-19588-avoid-poll-event

kirktrue committed:
Updated AbstractMembershipManager comments related to use of SharedAutoCommitState and SharedReconciliationState

kirktrue committed:
Merge branch 'trunk' into KAFKA-19588-avoid-poll-event

kirktrue committed:
s/Shared/ThreadSafe/g

Renames and refactors SharedConsumerState, SharedAutoCommitState, and SharedReconciliationState to ThreadSafe* equivalents. Introduces ThreadSafeConsumerState and ThreadSafeShareConsumerState abstractions. Updates all usages, constructors, and tests to use the new thread-safe classes, improving clarity and consistency in shared state management between application and network threads.

kirktrue committed:
Refactor auto-commit check in consumer state

Replaces the isAutoCommitEnabled() method in ThreadSafeAsyncConsumerState with direct access to autoCommitState().isAutoCommitEnabled(). Removes the redundant method and updates AsyncKafkaConsumer to use the new approach.

kirktrue committed:
Reverted unnecessary changes

kirktrue committed:
Refactor ThreadSafeAutoCommitState and ThreadSafeConsumerState

kirktrue committed:
Merge branch 'trunk' into KAFKA-19588-avoid-poll-event

kirktrue commented:
I'm closing this PR as we have decided against going any further with this approach due to thread safety concerns.

cc @lianetm @AndrewJSchofield
